lib/ts/controllers/s-popover.ts

```diff
@@ -275,6 +284,162 @@ namespace Stacks {
+        const referenceElement = referenceSelector ? element.querySelector(referenceSelector) : element;
+        const popoverId = referenceElement ? referenceElement.getAttribute("aria-controls") : null;
+        const popover = popoverId ? document.getElementById(popoverId) : null;
+        return [isPopover, controller, referenceElement, popover];
+    }
+
+    /**
+     * Helper to manually show an s-popover element via external JS
+     * @param element the element the `data-controller="s-popover"` attribute is on
+     */
+    export function showPopover(element: Element) {
+        const [isPopover, controller, _] = getPopover(element);
+
+        if (controller) {
+            controller.show();
+        } else if (isPopover) {
+            element.setAttribute("data-s-popover-show-on-connect", "true");
+        } else {
+            throw `element does not have data-controller="s-popover"`;
+        }
+    }
+
+    /**
+     * Helper to manually hide an s-popover element via external JS
+     * @param element the element the `data-controller="s-popover"` attribute is on
+     */
+    export function hidePopover(element: Element) {
+        const [isPopover, controller, _, popover] = getPopover(element);
+
+        if (controller) {
+            controller.hide();
+        } else if (isPopover) {
+            element.removeAttribute("data-s-popover-show-on-connect");
+            if (popover) {
+                popover.classList.remove("is-visible");
+            }
+        } else {
+            throw `element does not have data-controller="s-popover"`;
+        }
+    }
+
+    /**
+     * Options to use when attaching a popover via `Stacks.attachPopover`.
+     * @see Stacks.attachPopover
+     */
+    export interface PopoverOptions {
+        /**
+         * When true, the `click->s-popover#toggle` action will be attached to the controller element or reference element.
+         */
+        toggleOnClick?: boolean;
+        /**
+         * When set, `data-s-popover-placement` will be set to this value on the controller element.
+         */
+        placement?: string;
+
+        /**
+         * When true, the popover will appear immediately when the controller connects.
+         */
+        showOnConnect?: boolean;
+    }
+
+    /**
+     * Attaches a popover to an element and performs additional configuration.
+     * @param element the element that will receive the `data-controller="s-popover"` attribute.
+     * @param popover an element with the `.s-popover` class or HTML string containing a single element with the `.s-popover` class.
+     *                If the popover does not have a parent element, it will be inserted as a immediately after the reference element.
+     * @param options an optional collection of options to use when configuring the popover.
+     */
+    export function attachPopover(element: Element, popover: Element | string, options?: PopoverOptions)
```

Comment:
Is this going to be used to attach attributes to existing markup or to inject a popup into the dom as a string? Stimulus does dom mutation observation, so if you're adding in a popover via an ajax call (or similar), then you'd just render the attributes into the string as usual and let Stimulus hook up the element on insertion. No interference needed.

Can we hang sometime to go over this? I'm having trouble envisioning how this will be used.

Reply:
The goal is to go from having the following in the DOM:

```html
<button>React</button>
```

And the following payload from the server:

```
<div class="s-popover">...</div>
```

To having the following in the DOM:

```
<button data-controller="s-popover" data-action="s-popover#toggle" aria-controls="{uniqueId}">React</button>
<div class="s-popover is-visible" id="{uniqueId}">...</div>
```

Stimulus is smart enough that if the payload were the final state and I simply replaced the button with it, it would wire it up and work.  Unfortunately, in practice what we end up with is what I showed above, a button on the page and a popover in a JSON response.

The demo shows exactly what I want to do but for a real-world example, look at:

https://gh.stackoverflow.com/Core/StackOverflow/blob/master/StackOverflow/Content/Js/PartialJS/full/04_Vote.js#L908

That function could be reduced to:

```javascript
loadPopover() {

    if (Stacks.getControllerList(this.buttonTarget).contains("s-popover")) {
        return;
    }

    loadTemplate().done(html => {
        Stacks.attachPopover(this.button, html, { showOnConnect: true, placement: 'right", toggleOnClick: true });
        this.updateUI();
    });
}
```

Reply:
I'd like to improve this approach later, but I'll approve for now so we can get this into the next release. It should be fairly simple to clean up any usages if we decide to pivot later.